ligo/main.mligo

```diff
@@ -207,21 +225,118 @@ let set_position (s : storage) (i_l : tick_index) (i_u : tick_index) (i_l_l : ti
       }
     )
 
-let get_time_weighted_sum (s : storage) (c : views contract) : result =
-    ([Tezos.transaction (IC_sum s.time_weighted_ic_sum) 0mutez c], s)
+// Allocate slots in a buffer in [start, to_stop) indices range.
+// These slots will be filled with dummy values.
+let rec allocate_buffer_slots (buffer, start, to_stop : timed_cumulatives_buffer * nat * nat) : timed_cumulatives_buffer =
+    if start >= to_stop
+    then buffer
+    else
+        let new_buffer = {buffer with map = Big_map.add start dummy_timed_cumulatives buffer.map}
+        in allocate_buffer_slots(new_buffer, start + 1n, to_stop)
```

Comment:
I didn't actually try this, so please check if it makes sense in practice.

I think this has a couple of hidden costs:
1. due to how recursivity works in ligo the `start` and `to_stop` are to be paired and unpaired at each iteration, I suspect that this might cost more than perhaps creating and consuming a list
2. AFAIK using `dummy_timed_cumulatives` in the "loop" means that at each iteration a new "empty" value will be initialized, even tho this is in fact always the same and we could duplicate it

So what I'm thinking is to change the `increase_observation_count` to:
- use a small recursive function to create a list of the indices that we need to add
- initialize a constant with the value of `dummy_timed_cumulatives`
- `List.fold` over the list with the `big_map` as an accumulator

It _might_ end up being less expensive.

Reply:
1. How do you propose to construct the list? I seem to get the very same function of 3 arguments (index, termination index and accumulator).
2. That's hilarious, but I don't know how to ask LIGO to extract this constant :thinking:  If it is used once, then it is inlined, even if I assign it to a local variable and use this local variable.

    Indeed, I can pass the value through the loop. but this will slightly increase the contract size.

A bit funny which optimization issues we encounter here, and that cannot be workarounded, at least with Ocaml dilatect. Another issue I found when working in this your comment: [such one](https://gitlab.com/ligolang/ligo/-/issues/1279).

Reply:
Regarding n.1 I knew it would happen, but I was hoping that the cost of the recursive function to create the list would be smaller than the having the recursive `allocate_buffer_slots`.

BTW I did some tests and in terms of contract size I am right, it is less expensive to create the list and then `List.fold` over it than to recursively update the `big_map` directly.
However, the gain is not much and gas cost is increased, as basically two loops will be executed.
So that doesn't seem like a good idea :disappointed: 

I experimented with your current implementation a bit more and got it down by about `150` bytes, by rearranging some stuff (and avoiding the check at the beginning):
```ocaml
let increase_observation_count (s, new_val : storage * nat) : result =
    let rec allocate_buffer_slots (last_idx, b_map : nat * (nat, timed_cumulatives) big_map) : (nat, timed_cumulatives) big_map =
        let idx = last_idx + 1n in
        if idx >= new_val
        then b_map
        else allocate_buffer_slots (idx, Big_map.add idx dummy_timed_cumulatives b_map)
    in
    let buffer = s.cumulatives_buffer in
    let buffer_map = allocate_buffer_slots (buffer.reserved_length, buffer.map) in
    let new_buffer = { buffer with reserved_length = new_val; map = buffer_map }
    in (([] : operation list), {s with cumulatives_buffer = new_buffer})
```

Reply:
Mm, quite hilarious that even flipping the field assignments in `buffer` record changes the contract size by 8. Some precise fine tuning is still necessary.

Reply:
Yeah, however I'd leave such small improvements for later, it's not very useful to focus on them now IMO